File: cartography/fred.go
package cartography

import (
	"context"
	"encoding/json"
	"fmt"
	"math"
	"net/http"
	"net/url"
	"strconv"
	"time"
)

// FRED (Federal Reserve Economic Data) integration — pulls four canonical
// series and reduces them to a small set of binary recession/stress signals
// that override the formula's regime when reality disagrees with the model.
//
//   UNRATE        monthly unemployment rate           → Sahm rule
//   T10Y2Y        10Y minus 2Y Treasury spread, daily → yield-curve inversion
//   NFCI          Chicago Fed financial conditions    → credit/liquidity stress
//   BAMLH0A0HYM2  ICE BofA high-yield OAS, daily      → corporate stress

const fredAPIBase = "https://api.stlouisfed.org/fred/series/observations"

type fredObservation struct {
	Date  string `json:"date"`
	Value string `json:"value"` // FRED returns "." for missing
}

type fredResponse struct {
	Observations []fredObservation `json:"observations"`
}

// FREDClient hits the FRED API. Set APIKey from the FRED_API_KEY env var.
type FREDClient struct {
	APIKey string
	HTTP   *http.Client
}

// NewFREDClient builds a client with a sensible default HTTP timeout.
func NewFREDClient(apiKey string) *FREDClient {
	return &FREDClient{
		APIKey: apiKey,
		HTTP:   &http.Client{Timeout: 15 * time.Second},
	}
}

// fetch pulls the most recent `limit` observations of seriesID, sorted
// newest-first.
func (c *FREDClient) fetch(ctx context.Context, seriesID string, limit int) ([]fredObservation, error) {
	if c == nil || c.APIKey == "" {
		return nil, fmt.Errorf("FRED API key not configured")
	}
	q := url.Values{}
	q.Set("series_id", seriesID)
	q.Set("api_key", c.APIKey)
	q.Set("file_type", "json")
	q.Set("sort_order", "desc")
	q.Set("limit", strconv.Itoa(limit))

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, fredAPIBase+"?"+q.Encode(), nil)
	if err != nil {
		return nil, err
	}
	resp, err := c.HTTP.Do(req)
	if err != nil {
		return nil, fmt.Errorf("FRED %s: %w", seriesID, err)
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("FRED %s: HTTP %d", seriesID, resp.StatusCode)
	}
	var fr fredResponse
	if err := json.NewDecoder(resp.Body).Decode(&fr); err != nil {
		return nil, fmt.Errorf("FRED %s decode: %w", seriesID, err)
	}
	return fr.Observations, nil
}

// parseObs converts FRED observations to (values, dates), dropping "."
// missing markers. Output preserves desc-by-date order.
func parseObs(obs []fredObservation) ([]float64, []time.Time) {
	vals := make([]float64, 0, len(obs))
	dates := make([]time.Time, 0, len(obs))
	for _, o := range obs {
		if o.Value == "." || o.Value == "" {
			continue
		}
		v, err := strconv.ParseFloat(o.Value, 64)
		if err != nil {
			continue
		}
		d, err := time.Parse("2006-01-02", o.Date)
		if err != nil {
			continue
		}
		vals = append(vals, v)
		dates = append(dates, d)
	}
	return vals, dates
}

// computeSahm implements the Sahm Recession Indicator:
//   sahm = (current 3-month MA of UNRATE) - (lowest 3-month MA over prior 12 months)
// triggers at >= 0.50. Authoritative through the post-WWII record.
func computeSahm(obs []fredObservation) *DataSignal {
	vals, dates := parseObs(obs)
	if len(vals) < 14 {
		return nil
	}
	current := (vals[0] + vals[1] + vals[2]) / 3.0
	minMMA := math.Inf(1)
	// Window i=1..12: 3MMA at offset i uses vals[i], vals[i+1], vals[i+2].
	for i := 1; i <= 12 && i+2 < len(vals); i++ {
		mma := (vals[i] + vals[i+1] + vals[i+2]) / 3.0
		if mma < minMMA {
			minMMA = mma
		}
	}
	if math.IsInf(minMMA, 1) {
		return nil
	}
	delta := current - minMMA
	return &DataSignal{
		Source:    "FRED:UNRATE",
		Name:      "Sahm Recession Indicator",
		Value:     round2(delta),
		Threshold: 0.50,
		Triggered: delta >= 0.50,
		AsOf:      dates[0],
		Description: fmt.Sprintf("3M unemployment %.2f%% vs trailing-12mo min %.2f%% (Δ %+.2f, triggers ≥ +0.50)",
			current, minMMA, delta),
	}
}

// computeYieldCurve reads T10Y2Y. Negative = inverted (recession warning,
// historically 6–18 months lead).
func computeYieldCurve(obs []fredObservation) *DataSignal {
	vals, dates := parseObs(obs)
	if len(vals) == 0 {
		return nil
	}
	v := vals[0]
	state := "normal slope"
	if v < 0 {
		state = "INVERTED — recession-leading signal"
	} else if v < 0.25 {
		state = "near-flat"
	}
	return &DataSignal{
		Source:    "FRED:T10Y2Y",
		Name:      "10Y-2Y Treasury Spread",
		Value:     round2(v),
		Threshold: 0.0,
		Triggered: v < 0,
		AsOf:      dates[0],
		Description: fmt.Sprintf("10Y minus 2Y = %+.2f%% — %s", v, state),
	}
}

// computeNFCI reads the Chicago Fed National Financial Conditions Index.
// Mean = 0; positive = tighter than average. >0 already flags tightening.
func computeNFCI(obs []fredObservation) *DataSignal {
	vals, dates := parseObs(obs)
	if len(vals) == 0 {
		return nil
	}
	v := vals[0]
	state := "looser than average"
	if v > 0 {
		state = "tighter than average"
	}
	if v > 0.5 {
		state = "STRESSED — credit tightening"
	}
	return &DataSignal{
		Source:    "FRED:NFCI",
		Name:      "Financial Conditions (NFCI)",
		Value:     round2(v),
		Threshold: 0.0,
		Triggered: v > 0,
		AsOf:      dates[0],
		Description: fmt.Sprintf("NFCI = %+.2fσ — %s", v, state),
	}
}

// computeHYSpread reads the ICE BofA US High Yield OAS. Long-run median
// ~4–5%; sustained >7% historically marks credit-cycle stress.
func computeHYSpread(obs []fredObservation) *DataSignal {
	vals, dates := parseObs(obs)
	if len(vals) == 0 {
		return nil
	}
	v := vals[0]
	state := "calm"
	if v >= 5.5 {
		state = "elevated"
	}
	if v >= 7.0 {
		state = "STRESSED — credit-cycle warning"
	}
	return &DataSignal{
		Source:    "FRED:BAMLH0A0HYM2",
		Name:      "High-Yield Credit Spread (OAS)",
		Value:     round2(v),
		Threshold: 7.0,
		Triggered: v >= 7.0,
		AsOf:      dates[0],
		Description: fmt.Sprintf("HY OAS = %.2f%% — %s", v, state),
	}
}

// Snapshot fetches all four series and returns a composed feed. Failures of
// individual series are non-fatal — the snapshot returns the signals it
// could compute and surfaces the rest as a warning. Only a hard FRED error
// (no key, network down) returns an error.
func (c *FREDClient) Snapshot(ctx context.Context) (*DataFeed, error) {
	if c == nil || c.APIKey == "" {
		return nil, fmt.Errorf("FRED API key not configured")
	}

	feed := &DataFeed{LastFetched: time.Now()}

	// UNRATE: need 24 monthly obs for the trailing 12mo Sahm window.
	if obs, err := c.fetch(ctx, "UNRATE", 24); err == nil {
		if sig := computeSahm(obs); sig != nil {
			feed.Signals = append(feed.Signals, *sig)
		}
	} else {
		feed.Warnings = append(feed.Warnings, err.Error())
	}

	if obs, err := c.fetch(ctx, "T10Y2Y", 5); err == nil {
		if sig := computeYieldCurve(obs); sig != nil {
			feed.Signals = append(feed.Signals, *sig)
		}
	} else {
		feed.Warnings = append(feed.Warnings, err.Error())
	}

	if obs, err := c.fetch(ctx, "NFCI", 5); err == nil {
		if sig := computeNFCI(obs); sig != nil {
			feed.Signals = append(feed.Signals, *sig)
		}
	} else {
		feed.Warnings = append(feed.Warnings, err.Error())
	}

	if obs, err := c.fetch(ctx, "BAMLH0A0HYM2", 5); err == nil {
		if sig := computeHYSpread(obs); sig != nil {
			feed.Signals = append(feed.Signals, *sig)
		}
	} else {
		feed.Warnings = append(feed.Warnings, err.Error())
	}

	feed.Multiplier, feed.Triggers = composeMultiplier(feed.Signals)
	return feed, nil
}

File: rust-trader/crates/cartography/src/fred.rs
//! FRED client for economic data signals.

use chrono::{NaiveDate, Utc};
use serde::Deserialize;

use crate::feed::{compose_multiplier, DataFeed, DataSignal};

#[derive(Debug, Clone)]
pub struct FREDClient {
    api_key: String,
    http: reqwest::Client,
}

type SignalComputer = fn(&[f64], &[NaiveDate]) -> Option<DataSignal>;

#[derive(Deserialize)]
struct FredResp {
    observations: Vec<FredObs>,
}
#[derive(Deserialize)]
struct FredObs {
    date: String,
    value: String,
}

impl FREDClient {
    pub fn new(api_key: impl Into<String>) -> Self {
        Self {
            api_key: api_key.into(),
            http: reqwest::Client::new(),
        }
    }

    async fn fetch_observations(
        &self,
        series_id: &str,
        limit: usize,
    ) -> Result<Vec<FredObs>, Box<dyn std::error::Error + Send + Sync>> {
        let resp: FredResp = self
            .http
            .get("https://api.stlouisfed.org/fred/series/observations")
            .query(&[
                ("series_id", series_id),
                ("api_key", self.api_key.as_str()),
                ("file_type", "json"),
                ("sort_order", "desc"),
                ("limit", &limit.to_string()),
            ])
            .send()
            .await?
            .json()
            .await?;
        Ok(resp.observations)
    }

    fn parse_vals_dates(obs: &[FredObs]) -> (Vec<f64>, Vec<NaiveDate>) {
        obs.iter()
            .filter_map(|o| {
                if o.value == "." || o.value.is_empty() {
                    return None;
                }
                let v: f64 = o.value.parse().ok()?;
                let d = NaiveDate::parse_from_str(&o.date, "%Y-%m-%d").ok()?;
                Some((v, d))
            })
            .unzip()
    }

    fn sahm(vals: &[f64], dates: &[NaiveDate]) -> Option<DataSignal> {
        if vals.len() < 14 {
            return None;
        }
        let cur = (vals[0] + vals[1] + vals[2]) / 3.0;
        let mut min_mma = f64::INFINITY;
        for i in 1..=12 {
            if i + 2 >= vals.len() {
                break;
            }
            let mma = (vals[i] + vals[i + 1] + vals[i + 2]) / 3.0;
            if mma < min_mma {
                min_mma = mma;
            }
        }
        let delta = cur - min_mma;
        let as_of = dates.first()?.and_hms_opt(0, 0, 0)?.and_utc();
        Some(DataSignal {
            source: "FRED:UNRATE".into(),
            name: "Sahm Recession Indicator".into(),
            value: crate::round2(delta),
            threshold: 0.50,
            triggered: delta >= 0.50,
            as_of,
            description: format!("3M MA {:.2}% vs min {:.2}% (d={:+.2})", cur, min_mma, delta),
        })
    }

    fn yield_curve(vals: &[f64], dates: &[NaiveDate]) -> Option<DataSignal> {
        let v = vals.first()?;
        let as_of = dates.first()?.and_hms_opt(0, 0, 0)?.and_utc();
        Some(DataSignal {
            source: "FRED:T10Y2Y".into(),
            name: "10Y-2Y Treasury Spread".into(),
            value: crate::round2(*v),
            threshold: 0.0,
            triggered: *v < 0.0,
            as_of,
            description: format!("10Y-2Y = {:+.2}%", v),
        })
    }

    fn nfci(vals: &[f64], dates: &[NaiveDate]) -> Option<DataSignal> {
        let v = vals.first()?;
        let as_of = dates.first()?.and_hms_opt(0, 0, 0)?.and_utc();
        Some(DataSignal {
            source: "FRED:NFCI".into(),
            name: "Financial Conditions (NFCI)".into(),
            value: crate::round2(*v),
            threshold: 0.0,
            triggered: *v > 0.0,
            as_of,
            description: format!("NFCI = {:+.2}", v),
        })
    }

    fn hy_spread(vals: &[f64], dates: &[NaiveDate]) -> Option<DataSignal> {
        let v = vals.first()?;
        let as_of = dates.first()?.and_hms_opt(0, 0, 0)?.and_utc();
        Some(DataSignal {
            source: "FRED:BAMLH0A0HYM2".into(),
            name: "High-Yield Credit Spread (OAS)".into(),
            value: crate::round2(*v),
            threshold: 7.0,
            triggered: *v >= 7.0,
            as_of,
            description: format!("HY OAS = {:.2}%", v),
        })
    }

    fn ig_spread(vals: &[f64], dates: &[NaiveDate]) -> Option<DataSignal> {
        let v = vals.first()?;
        let as_of = dates.first()?.and_hms_opt(0, 0, 0)?.and_utc();
        Some(DataSignal {
            source: "FRED:BAMLC0A0CM".into(),
            name: "Investment-Grade Credit Spread (OAS)".into(),
            value: crate::round2(*v),
            threshold: 2.0,
            triggered: *v >= 2.0,
            as_of,
            description: format!("IG corporate OAS = {:.2}%", v),
        })
    }

    pub async fn snapshot(&self) -> Result<DataFeed, Box<dyn std::error::Error + Send + Sync>> {
        let mut feed = DataFeed {
            signals: Vec::new(),
            triggers: Vec::new(),
            multiplier: 1.0,
            warnings: Vec::new(),
            last_fetched: Utc::now(),
        };

        let series: &[(&str, usize, SignalComputer)] = &[
            ("UNRATE", 24, Self::sahm),
            ("T10Y2Y", 5, Self::yield_curve),
            ("NFCI", 5, Self::nfci),
            ("BAMLH0A0HYM2", 5, Self::hy_spread),
            ("BAMLC0A0CM", 5, Self::ig_spread),
        ];

        for (sid, limit, compute) in series {
            match self.fetch_observations(sid, *limit).await {
                Ok(obs) => {
                    let (vals, dates) = Self::parse_vals_dates(&obs);
                    if let Some(sig) = compute(&vals, &dates) {
                        feed.signals.push(sig);
                    }
                }
                Err(e) => feed.warnings.push(format!("{}: {}", sid, e)),
            }
        }

        (feed.multiplier, feed.triggers) = compose_multiplier(&feed.signals);
        Ok(feed)
    }
}
